Woodoku/tests/entity/test_woodoku_board.py

```diff
@@ -0,0 +1,163 @@
+from typing import Tuple
+
+import numpy as np
+import pytest
+
+from exceptions.exceptions import ShapeOutOfBoardError
+from woodoku_board import WoodokuBoard
+from woodoku_shape import WoodokuShape
+
+
+class TestWoodokuBoard:
+    l_shape: WoodokuShape = WoodokuShape([(0, 0), (1, 0), (1, 1), (1, 2)])
+    horizontal_bar_shape: WoodokuShape = WoodokuShape([(0, 0), (0, 1), (0, 2), (0, 3), (0, 4)])
+    gun_shape: WoodokuShape = WoodokuShape([(0, 0), (0, 1), (0, 2), (1, 0)])
+    vertical_two_block: WoodokuShape = WoodokuShape([(0, 0), (1, 0)])
+    one_block: WoodokuShape = WoodokuShape([(0, 0)])
+
+    @pytest.mark.parametrize('shape', [
+        l_shape,
+        horizontal_bar_shape,
+        gun_shape
+    ])
+    def test_add_shape_on_empty_board(self, shape: WoodokuShape) -> None:
+        """
+        Add shape to valid position on the board and remove all blocks of the shape to test add_shape
+        """
+        board = WoodokuBoard()
+        board.add_shape(shape, 5, 0)
+        board._WoodokuBoard__representation.remove_blocks(shape.map_to_board_at(5, 0))
+        assert (board._WoodokuBoard__representation._WoodokuBoardRepresentation__board == np.full((9, 9), False)).all()
+
+    @pytest.mark.parametrize('first_shape, first_position, sec_shape, sec_position', [
+        (l_shape, (0, 0), gun_shape, (4, 0)),
+        (horizontal_bar_shape, (1, 0), gun_shape, (2, 0)),
+        (gun_shape, (3, 0), horizontal_bar_shape, (1, 0))
+    ])
+    def test_add_shape_on_board_without_conflict(self, first_shape: WoodokuShape, first_position: Tuple[int, int],
+                                                 sec_shape: WoodokuShape, sec_position: Tuple[int, int]) -> None:
+        """
+        first_shape and first_location are chosen so that sec_shape will be NOT overlapped with first_shape when
+        adding it at sec_location
+        """
+        board = WoodokuBoard()
+        try:
+            board.add_shape(first_shape, *first_position)
+            board.add_shape(sec_shape, *sec_position)
+        except ShapeOutOfBoardError as exc:
+            assert False, f"shape with coordinates: {first_shape.get_coords} , {sec_shape} " \
+                          f"and positions: {first_position}, {sec_position} raised an exception {exc}"
```

Comment:
There is an error in representation, the only method __validate() which throws a error is never used

Reply:
Yes, the test case passes when no error is thrown. I hardly recall my intention to catch the error here, but we could raise an exception in add_shape to reinforce the precondition.

Reply:
We shouldn't raise error in `add_shape` since we only add shapes when we know we can add it that is the precondition. 
The error should be raised in the representation using `__valid` method